tortoise/tortoise.go

```diff
@@ -636,268 +449,264 @@ func (t *turtle) getBallotBeacon(ballot *types.Ballot, logger log.Log) ([]byte,
-func (t *turtle) verifyLayers(ctx *tcontext) error {
-	target := t.Processed
-	logger := t.logger.WithContext(ctx).WithFields(
-		log.Named("verification_target", target),
-		log.Named("old_verified", t.Verified))
-
-	// attempt to verify each layer from the last verified up to one prior to the newly-arrived layer.
-	// this is the full range of unverified layers that we might possibly be able to verify at this point.
-	// Note: t.Verified is initialized to the effective genesis layer, so the first candidate layer here necessarily
-	// follows and is post-genesis. There's no need for an additional check here.
-	for candlid := t.Verified.Add(1); candlid.Before(target); candlid = candlid.Add(1) {
-		logger := logger.WithFields(log.Named("candidate_layer", candlid))
-
-		// it's possible that self healing already verified a layer
-		if !t.Verified.Before(candlid) {
-			logger.Info("self healing already verified this layer")
-			continue
+	for _, part := range [...][]*types.Ballot{refs, other} {
+		for _, ballot := range part {
+			ballotWeight, err := computeBallotWeight(t.atxdb, t.ballotWeight, ballot, t.LayerSize, types.GetLayersPerEpoch())
+			if err != nil {
+				return nil, err
+			}
+			t.ballotWeight[ballot.ID()] = ballotWeight
+			t.ballotLayer[ballot.ID()] = ballot.LayerIndex
+
+			// TODO(dshulyak) this should not fail without terminating tortoise
+			t.ballotHasGoodBeacon(ballot, t.logger)
+
+			ballotsIDs = append(ballotsIDs, ballot.ID())
+
+			baselid := t.ballotLayer[ballot.BaseBallot]
+			if baselid.Before(min) {
+				min = baselid
+			}
+			votes := Opinion{}
+			for lid := baselid; lid.Before(processed); lid = lid.Add(1) {
+				for _, bid := range t.blocks[lid] {
+					votes[bid] = against
+				}
+			}
+			for _, bid := range ballot.ForDiff {
+				votes[bid] = support
+			}
+			for _, bid := range ballot.NeutralDiff {
+				votes[bid] = abstain
+			}
+			for _, bid := range ballot.AgainstDiff {
+				votes[bid] = against
+			}
+			ballots = append(ballots, tortoiseBallot{
+				id:     ballot.ID(),
+				base:   ballot.BaseBallot,
+				weight: ballotWeight,
+				votes:  votes,
+			})
 		}
+	}
+	t.ballots[lid] = ballotsIDs
+	return ballots, nil
+}
 
-		logger.Info("attempting to verify candidate layer")
-		contextualValidity, err := t.verifyingTortoise(ctx, logger, candlid)
-		if err != nil {
+func (t *turtle) updateLocalOpinion(ctx *tcontext, from, to types.LayerID) error {
+	for lid := from; !lid.After(to); lid = lid.Add(1) {
+		if err := t.addLocalOpinion(ctx, lid, t.localOpinion); err != nil {
 			return err
 		}
+	}
+	return nil
+}
 
-		if contextualValidity != nil {
-			for blk, v := range contextualValidity {
-				if err := t.bdp.SaveContextualValidity(blk, candlid, v); err != nil {
-					logger.With().Error("error saving contextual validity on block", blk, log.Err(err))
-					return fmt.Errorf("saving contextual validity for %s: %w", blk, err)
-				}
+func (t *turtle) keepVotes(ballots []tortoiseBallot) {
```

Comment:
i wonder whether we can keep a separate eviction policy for base ballots. most ballot in a layer should have the same base ballot. and we only evict base ballots when there is no references to it anymore.

Reply:
i intend to change the way we store votes in https://github.com/spacemeshos/go-spacemesh/pull/2994